src/cpp/fastdds/xtypes/type_representation/TypeObjectUtils.cpp

```diff
@@ -0,0 +1,3017 @@
+            break;
+
+        // Primitive TypeIdentifiers/ExtendedTypeDefn/EquivalenceHash: no inconsistency rule apply.
+        case EK_COMPLETE:
+        case EK_MINIMAL:
+        default:
+            break;
+    }
+}
+
+void TypeObjectUtils::applied_annotation_parameter_seq_consistency(
+        const AppliedAnnotationParameterSeq& applied_annotation_parameter_seq)
+{
+    std::set<NameHash> param_hashes;
+    for (size_t i = 0; i < applied_annotation_parameter_seq.size(); i++)
+    {
+        param_hashes.insert(applied_annotation_parameter_seq[i].paramname_hash());
+        if (param_hashes.size() != (i + 1))
+        {
+            throw InvalidArgumentError("Repeated annotation parameters in the sequence");
+        }
+    }
+}
+
+void TypeObjectUtils::applied_annotation_type_identifier_consistency(
+        const TypeIdentifier& annotation_type_id)
+{
+    TypeObject type_object;
+    ReturnCode_t ret_code = DomainParticipantFactory::get_instance()->type_object_registry().get_type_object(
+        annotation_type_id, type_object);
+    if (ReturnCode_t::RETCODE_OK == ret_code)
+    {
+        if ((EK_COMPLETE == type_object._d() && type_object.complete()._d() != TK_ANNOTATION) ||
+                (EK_MINIMAL == type_object._d() && type_object.minimal()._d() != TK_ANNOTATION))
+        {
+            throw InvalidArgumentError("AppliedAnnotation TypeIdentifier does not correspond with an Annotation type");
+        }
+    }
+    else if (ReturnCode_t::RETCODE_NO_DATA == ret_code)
+    {
+        throw InvalidArgumentError("Annotation TypeIdentifier unknown to TypeObjectRegistry");
+    }
+    else
+    {
+        throw InvalidArgumentError("Annotation TypeIdentifier is not direct HASH");
+    }
+}
+
+void TypeObjectUtils::applied_annotation_consistency(
+        const AppliedAnnotation& applied_annotation)
+{
+    applied_annotation_type_identifier_consistency(applied_annotation.annotation_typeid());
+    if (applied_annotation.param_seq().has_value())
+    {
+        applied_annotation_parameter_seq_consistency(applied_annotation.param_seq().value());
+    }
+    if (DomainParticipantFactory::get_instance()->type_object_registry().is_builtin_annotation(
+                applied_annotation.annotation_typeid()))
+    {
+        throw InvalidArgumentError("Builtin annotation cannot be defined as custom annotation");
+    }
+}
+
+void TypeObjectUtils::applied_annotation_seq_consistency(
+        const AppliedAnnotationSeq& applied_annotation_seq)
+{
+    std::unordered_set<TypeIdentifier> annotation_typeids;
+    for (size_t i = 0; i < applied_annotation_seq.size(); i++)
+    {
+        annotation_typeids.insert(applied_annotation_seq[i].annotation_typeid());
+        if (annotation_typeids.size() != (i + 1))
+        {
+            throw InvalidArgumentError("Repeated annotation in the sequence");
+        }
+        applied_annotation_consistency(applied_annotation_seq[i]);
+    }
```

Comment:
```suggestion
    for (const auto& parameter : applied_annotation_parameter_seq)
    {
        if (!param_hashes.insert(parameter.paramname_hash()).second)
        {
            throw InvalidArgumentError("Repeated annotation parameters in the sequence");
        }
        applied_annotation_consistency(applied_annotation_seq[i]);
    }
```
The insert function returns a pair with a boolean indicating if the insertion was successful.